loaders/src/glTF/2.0/Extensions/KHR_materials_variants.ts

```diff
@@ -0,0 +1,132 @@
+import { Nullable } from "babylonjs/types";
+import { IGLTFLoaderExtension } from "../glTFLoaderExtension";
+import { GLTFLoader, ArrayItem } from "../glTFLoader";
+
+import { Material } from 'babylonjs/Materials/material';
+import { Mesh } from 'babylonjs/Meshes/mesh';
+import { AbstractMesh } from 'babylonjs/Meshes/abstractMesh';
+import { INode, IMeshPrimitive, IMesh } from '../glTFLoaderInterfaces';
+
+const NAME = "KHR_materials_variants";
+
+interface IKHRMaterialVariantsMapping {
+    tags: string[];
+    material: number;
+}
+
+interface IKHRMaterialVariants {
+    mapping: IKHRMaterialVariantsMapping[];
+}
+
+interface IKHRMaterialVariantsTop {
+    default?: string;
+}
+
+interface VariantMapping {
+    mesh: AbstractMesh;
+    materialPromise: Promise<Nullable<Material>>;
+    material?: Material;
+}
+
+/**
+ * [Proposed Specification](https://github.com/KhronosGroup/glTF/pull/1681)
+ * !!! Experimental Extension Subject to Changes !!!
+ */
+export class KHR_materials_variants implements IGLTFLoaderExtension {
+    /**
+     * The name of this extension.
+     */
+    public readonly name = NAME;
+
+    /**
+     * Defines whether this extension is enabled.
+     */
+    public enabled: boolean;
+
+    private _loader: GLTFLoader;
+
+    public defaultVariant: string | undefined;
+    public availableVariants: string[];
+
+    private _tagsToMap: { [key: string]: VariantMapping[]; } = {};
+
+    constructor(loader: GLTFLoader) {
+        this._loader = loader;
+        this.enabled = this._loader.isExtensionUsed(NAME);
+    }
+
+    /** @hidden */
+    public dispose() {
+        delete this._loader;
+    }
+
+    public getVariants(): string[] {
+        return Object.keys(this._tagsToMap);
+    }
+
+    public selectVariant(variantName: string | string[]) {
```

Comment:
This is method and data is shared and accumulated by all models loaded. That would be ok for most of the cases of 3D Commerce where typically only one model is loaded during a session. But I believe it would be worth striving for support of multiple models loaded and allow managing each separately.

At the same time I am not sure about the appropriate mechanism:

1. Can we add metadata to the root that exposes the data and methods?
2. Can we return a custom wrapper that expose that?
3. Can we get an instance of the returned AssetContainer and use a WeakMap to link the data with that?
4. Can we just store the mappings in mesh metadata and allow users to instantiate a Manager for an AssetContainer that gathers the data and exposes the functionality?
5. Something else?

I am leaning towards 4 but I wonder what others think.

Reply:
> This is method and data is shared and accumulated by all models loaded.

This isn't true. Each time a glTF is loaded, new loader extensions are created. So, this is per glTF asset. @MiiBond and I are still discussing this offline on whether this is the right way to select variants though.